File: appactions/interaction/interaction-capabilities-core/src/main/java/androidx/appactions/interaction/capabilities/core/values/Thing.java
/*
 * Copyright 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.appactions.interaction.capabilities.core.values;

import androidx.annotation.NonNull;

import java.util.Optional;

/** Common interface for structured entity. */
public abstract class Thing {
    /** Returns the id of this thing. */
    @NonNull
    public abstract Optional<String> getId();

    /** Returns the name of this thing. */
    @NonNull
    public abstract Optional<String> getName();

    /**
     * Base builder class that can be extended to build objects that extend Thing.
     *
     * @param <T>
     */
    public abstract static class Builder<T extends Builder<T>> {
        /** Sets the id of the Thing to be built. */
        @NonNull
        public abstract T setId(@NonNull String id);

        /** Sets the name of the Thing to be built. */
        @NonNull
        public abstract T setName(@NonNull String name);
    }
}

File: appactions/interaction/interaction-capabilities-core/src/main/java/androidx/appactions/builtintypes/experimental/types/Thing.kt
/*
 * Copyright 2023 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package androidx.appactions.builtintypes.experimental.types

import androidx.appactions.builtintypes.experimental.properties.Name

interface Thing {
    val identifier: String?
    val name: Name?

    fun toBuilder(): Builder<*>

    companion object {
        @JvmStatic
        fun Builder(): Builder<*> = ThingBuilderImpl()
    }

    @Suppress("StaticFinalBuilder")
    interface Builder<Self : Builder<Self>> {
        fun build(): Thing

        fun setIdentifier(text: String?): Self
        fun setName(text: String): Self
        fun setName(name: Name?): Self
        fun clearName(): Self
    }
}

private class ThingBuilderImpl : Thing.Builder<ThingBuilderImpl> {

    private var identifier: String? = null
    private var name: Name? = null

    override fun build(): Thing = ThingImpl(identifier, name)

    override fun setIdentifier(text: String?): ThingBuilderImpl =
        apply { identifier = text }

    override fun setName(text: String): ThingBuilderImpl = apply { name = Name(text) }
    override fun setName(name: Name?): ThingBuilderImpl = apply { this.name = name }
    override fun clearName(): ThingBuilderImpl = apply { name = null }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true

        other as ThingBuilderImpl
        if (this.name != other.name) return false
        if (this.identifier != other.identifier) return false
        return true
    }
}

private class ThingImpl(override val identifier: String?, override val name: Name?) : Thing {
    override fun toBuilder(): Thing.Builder<*> =
        ThingBuilderImpl().setIdentifier(identifier).setName(name)
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ThingImpl) return false
        if (this.name != other.name) return false
        if (this.identifier != other.identifier) return false
        return true
    }
}